lib/MongoDB/_Protocol.pm

```diff
@@ -389,18 +533,18 @@ use constant {
 };
 
 sub parse_reply {
-    my ( $msg, $request_id ) = @_;
+    my ( $msg, $request_id, $codec ) = @_;
 
     MongoDB::ProtocolError->throw("response was truncated")
         if length($msg) < MIN_REPLY_LENGTH;
 
     $msg = try_uncompress($msg);
-
+    
     my (
         $len, $msg_id, $response_to, $opcode, $bitflags, $cursor_id, $starting_from,
         $number_returned
-    ) = unpack( P_REPLY_HEADER, $msg );
-
+    ) = unpack( P_MSG, $msg );
```

Comment:
I don't think this works because replies can be either OP_REPLY or OP_MSG. You need to unpack just the header and do opcode inspection early and take different paths.